src/practices/Rust/RustGitignoreCorrectlySetPractice.ts

```diff
@@ -0,0 +1,77 @@
+import { PracticeBase } from '../PracticeBase';
+
+import { PracticeEvaluationResult, PracticeImpact, ProgrammingLanguage } from '../../model';
+import { DxPractice } from '../DxPracticeDecorator';
+import { PracticeContext } from '../../contexts/practice/PracticeContext';
+import { ReportDetailType } from '../../reporters/ReporterData';
+import { FixerContext } from '../../contexts/fixer/FixerContext';
+import { IFileInspector } from '../../inspectors';
+
+@DxPractice({
+  id: 'Rust.GitignoreCorrectlySet',
+  name: 'Set .gitignore correctly',
+  impact: PracticeImpact.high,
+  suggestion: 'Set patterns in the .gitignore file as usual.',
+  reportOnlyOnce: true,
+  url: 'https://github.com/github/gitignore/blob/master/Rust.gitignore',
+  dependsOn: { practicing: ['LanguageIndependent.GitignoreIsPresent'] },
+})
+export class RustGitignoreCorrectlySetPractice extends PracticeBase {
+  private IGNORE_CHECKS: [RegExp, string][] = [
+    [/target\//, 'target/'],
+    [/\*\*\/\*\.rs\.bk/, '**/*.rs.bk'],
+  ];
+
+  async isApplicable(ctx: PracticeContext): Promise<boolean> {
+    return ctx.projectComponent.language === ProgrammingLanguage.Rust;
+  }
+
+  private static async parseGitignore(fileInspector: IFileInspector): Promise<{ raw: string; rules: string[] }> {
```

Comment:
Please, keep the same structure as we have in other practices for checking correctly set gitignore. E.g. C#, JavaScript, ... It's quite hard to check it if all the code is correct.

Reply:
I think this could warrant some kind of a base class that each language can extend for the gitignore practice specifically (and also other practices that need to be customizable to some degree as well).

For example, I can imagine something like this
```ts
abstract class GitignorePresentPracticeBase extends PracticeBase {
  protected constructor(
    protected readonly applicableLanguages: ProgrammingLanguage[],
    protected readonly ruleChecks: { check: RegExp, fix: string }[] 
  ) {
    super();
  }

  protected static async parseGitignore(fileInspector: IFileInspector): Promise<{ raw: string; rules: string[] }> {
    const raw = await fileInspector.readFile('.gitignore');
    const rules = raw.split('/\r?\n/').filter((line) => !line.startsWith('#') && line.trim() !== '');

    return {
      raw,
      rules,
    };
  }
  
  async isApplicable(ctx: PracticeContext): Promise<boolean> {
    return this.applicableLanguages.includes(ctx.projectComponent.language);
  }

  async evaluate(ctx: PracticeContext): Promise<PracticeEvaluationResult> {
    if (!ctx.root.fileInspector) {
      return PracticeEvaluationResult.unknown;
    }

    const gitignore = await GitignorePresentPracticeBase.parseGitignore(ctx.root.fileInspector);
    const allFound = this.ruleChecks.every((rule) => gitignore.rules.find((v) => rule.check.test(v)) !== undefined);

    if (allFound) {
      return PracticeEvaluationResult.practicing;
    }

    return PracticeEvaluationResult.notPracticing;
  }

  async fix(ctx: FixerContext) {
    const inspector = ctx.fileInspector?.basePath ? ctx.fileInspector : ctx.root.fileInspector;
    if (!inspector) return;

    const gitignore = await GitignorePresentPracticeBase.parseGitignore(inspector);
    const fixes = this.ruleChecks.filter((rule) => gitignore.rules.find((v) => rule.check.test(v)) === undefined).map(
      (rule) => rule.fix,
    );

    if (fixes.length > 0) {
      let fixesString = fixes.join('\n');
      if (fixesString.length > 0) {
        fixesString = '\n# added by `dx-scanner --fix`\n' + fixesString;
      }
      // TODO: Fix the race condition when two components are at the same path
      await inspector.writeFile('.gitignore', `${gitignore.raw}${fixesString}`);
    }
  }
}
```

then the rust practice checker would be much slimmer. In fact, it would factor a big chunk of common code out of `RubyGitignoreCorrectlySetPractice`, `JsGitignoreCorrectlySetPractice`, `CSharpGitignoreCorrectlySetPractice` and a moderate chunk out of `TsGitignoreCorrectlySetPractice` and `JavaGitignoreCorrectlySetPractice`.

```ts
@DxPractice({
  id: 'Rust.GitignoreCorrectlySet',
  name: 'Set .gitignore correctly',
  impact: PracticeImpact.high,
  suggestion: 'Set patterns in the .gitignore file as usual.',
  reportOnlyOnce: true,
  url: 'https://github.com/github/gitignore/blob/master/Rust.gitignore',
  dependsOn: { practicing: ['LanguageIndependent.GitignoreIsPresent'] },
})
export class RustGitignoreCorrectlySetPractice extends GitignorePresentPracticeBase {
  constructor() {
    super(
      [ProgrammingLanguage.Rust],
      [
        { check: /target\//, fix: 'target/' },
        { check: /\*\*\/\*\.rs\.bk/, fix: '**/*.rs.bk'}
      ]
    );
  }

  async evaluate(ctx: PracticeContext): Promise<PracticeEvaluationResult> {
    const result = await super.evaluate(ctx);

    if (result === PracticeEvaluationResult.notPracticing) {
      this.data.details = [
        {
          type: ReportDetailType.text,
          text: 'You should ignore the `target` folder, and for libraries also the `Cargo.lock` file.',
        },
      ];
    }
    return result;
  }
}
```

reads better, doesn't it?

Please consider. If you'd be open to this change, I'll open a separate PR which would also update all existing gitignore practices to use such base class.

Reply:
This is interesting idea. Problem can be with constructor. `GitignorePresentPracticeBase`has to be, from IOC point of view, just another practice. It means the it can access it's dependencies via `ctx: PracticeContext`. Also don't forget to call `this.setData` before `return PracticeEvaluationResult.notPracticing;`.

Reply:
`GitignorePresentPracticeBase` is abstract, same as `PracticeBase`, I'm not introducing any new concepts. The constructor doesn't matter because it's all static data in there anyway. It's just there to canonically pass information to the base class so that the implementations don't differ when they all do the same. Please realize that all the gitignore practices do exatly the same parsing and regex matching, and each of them repeats the same code. Only the Java and Typescript practices do some additional work afterwards, but they still go through `parseGitignore -> matchRegexes -> checkIfTheyAllMatch` and honestly they even do it pretty inefficiently at that. Having it in a common class should allow easier refactoring and better performance tweaking.

Additional;y, `this.setData` is actually not defined on `PracticeBase`. I see that is it an ad-hoc convention, but even between all the practices it's not consistent as sometimes [it is private,](https://github.com/DXHeroes/dx-scanner/blob/bf3cfacdcd149f16f591269806f275ce70e2ba3b/src/practices/CSharp/C%23GitignoreCorrectlySetPractice.ts#L83) sometimes [public and takes arguments](https://github.com/DXHeroes/dx-scanner/blob/bf3cfacdcd149f16f591269806f275ce70e2ba3b/src/practices/Java/JavaDependenciesVersionMajorLevel.ts#L58) and sometimes [it is inline](https://github.com/DXHeroes/dx-scanner/blob/bf3cfacdcd149f16f591269806f275ce70e2ba3b/src/practices/LanguageIndependent/CorrectCommitMessagesPractice.ts#L43).

That is exactly why in my comment I proposed to override `evaluate` method, call `super.evaluate()` and then set the data as I need it depending on the result. I think this is the exact problem inheritance is supposed to solve.

Reply:
Good point 👍 I am aware of repeated code that's why I like your idea. 🙂

Yes, it is an ad-hoc convention. 

Calling `setData` in child class is OK. My point was just not to forget calling it.

You can open separate PR for this change 🙂
I really like your approach 🤩